js/src/features/project/components/ProjectRoleForm.tsx

```diff
@@ -1,54 +1,67 @@
-import React, { useCallback, useEffect, useMemo, useState } from "react";
-import { Button, Form, Input, Checkbox } from "antd";
-import { EditOutlined, DeleteOutlined } from "@ant-design/icons";
+import React, { useEffect, useMemo, useRef, useState } from "react";
+import { Button, Input, Checkbox } from "antd";
+import {
+  EditOutlined,
+  DeleteOutlined,
+  CheckCircleOutlined,
+  CloseCircleOutlined,
+} from "@ant-design/icons";
 
-import { useFormServerErrors } from "hooks/useFormServerErrors";
-import { FormErrorMap } from "MyTypes";
 import {
   Project,
   EditProjectRoleRequest,
   RemoveProjectRoleRequest,
   CreateProjectRoleRequest,
+  RolePermissions,
 } from "ProjectModels";
+import { ProjectRoleFormTableWrapper } from "./ProjectRoleFormTableWrapper";
 
 import "./ProjectRoleForm.scss";
-import { ProjectRoleFormTableWrapper } from "./ProjectRoleFormTableWrapper";
 
 interface Props {
   project: Project;
   create: (form: CreateProjectRoleRequest) => void;
   edit: (form: EditProjectRoleRequest) => void;
   remove: (form: RemoveProjectRoleRequest) => void;
-  errors: FormErrorMap;
-  loading: boolean;
+  rolePermissions: RolePermissions;
+}
+
+interface FormField {
+  id: string;
+  name: string;
+  permissions: string[];
+}
+
+interface Form {
+  [key: string]: FormField;
 }
 
+const minLengthRoleName = 3;
+
 export const ProjectRoleForm: React.FC<Props> = ({
-  errors,
   create,
   edit,
   remove,
   project,
-  loading,
+  rolePermissions,
 }) => {
-  // TODO: ADD PERMISSIONS WHEN ADD ENDPOINT
-  const permissions = ["testing", "building", "scripting", "deploy"];
-  const [form] = Form.useForm();
-
-  const [permissionIndexField, setPermissionIndexField] = useState(-1);
-
-  const initialValues = useMemo(
-    () => ({
-      roles: project.roles,
-    }),
-    [project],
+  const initialForm = useMemo(
+    () =>
+      project.roles.reduce(
+        (acc, role, index) => ({
+          ...acc,
+          [index]: role,
+        }),
+        {},
+      ),
+    [project.roles],
   );
 
-  useFormServerErrors(form, errors, ["name", "description"]);
+  const [indexOfEditableInput, setIndexOfEditableInput] = useState(-1);
+  const [indexOfEditableField, setIndexOfEditableField] = useState(-1);
+  const [form, setForm] = useState<Form>(initialForm);
```

Comment:
Why did you decide to change Ant.design form to React.State ? I think you can make this form with ant. They have dynamic forms: https://ant.design/components/form/#components-form-demo-dynamic-form-item

Reply:
If you use a table, you cannot use Checkbox.Group to update the state of the form, because of Checkbox.Group wraps internal Checkbox in a div and because of this the layout of the table will go

Reply:
I am agree with @Aliaksandrzaz. This form actually is not a form and roles are not nested fields of big form :) This just a table with independent rows.

Reply:
If it is not a form why is it called `form` ?

Reply:
But it is form for one role. Independent.

Reply:
No, it is not. It's for all roles